File: back/src/main/java/com/back/global/security/oauth2/converter/CustomTokenResponseConverter.java
package com.back.global.security.oauth2.converter;

import org.springframework.core.convert.converter.Converter;
import org.springframework.security.oauth2.core.OAuth2AccessToken;
import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse;
import org.springframework.security.oauth2.core.endpoint.OAuth2ParameterNames;

import java.util.*;
import java.util.stream.Collectors;

public class CustomTokenResponseConverter implements Converter<Map<String, Object>, OAuth2AccessTokenResponse> {

    private static final Set<String> TOKEN_RESPONSE_PARAMETER_NAMES = Set.of(
            OAuth2ParameterNames.ACCESS_TOKEN,
            OAuth2ParameterNames.TOKEN_TYPE,
            OAuth2ParameterNames.EXPIRES_IN,
            OAuth2ParameterNames.REFRESH_TOKEN,
            OAuth2ParameterNames.SCOPE
    );

    @Override
    public OAuth2AccessTokenResponse convert(Map<String, Object> tokenResponseParameters) {
        String accessToken = (String) tokenResponseParameters.get(OAuth2ParameterNames.ACCESS_TOKEN);
        OAuth2AccessToken.TokenType tokenType = OAuth2AccessToken.TokenType.BEARER;

        long expiresIn = 0;
        if (tokenResponseParameters.containsKey(OAuth2ParameterNames.EXPIRES_IN)) {
            expiresIn = Long.parseLong(tokenResponseParameters.get(OAuth2ParameterNames.EXPIRES_IN).toString());
        }

        Set<String> scopes = Collections.emptySet();
        if (tokenResponseParameters.containsKey(OAuth2ParameterNames.SCOPE)) {
            String scope = (String) tokenResponseParameters.get(OAuth2ParameterNames.SCOPE);
            scopes = Arrays.stream(scope.split(" ")).collect(Collectors.toSet());
        }

        Map<String, Object> additionalParameters = new LinkedHashMap<>();
        tokenResponseParameters.forEach((key, value) -> {
            if (!TOKEN_RESPONSE_PARAMETER_NAMES.contains(key)) {
                additionalParameters.put(key, value);
            }
        });

        if (tokenResponseParameters.containsKey(OAuth2ParameterNames.REFRESH_TOKEN)) {
            additionalParameters.put(
                    OAuth2ParameterNames.REFRESH_TOKEN,
                    tokenResponseParameters.get(OAuth2ParameterNames.REFRESH_TOKEN)
            );
        }

        OAuth2AccessTokenResponse.Builder builder = OAuth2AccessTokenResponse.withToken(accessToken)
                .tokenType(tokenType)
                .expiresIn(expiresIn)
                .scopes(scopes)
                .additionalParameters(additionalParameters);

        if (tokenResponseParameters.containsKey(OAuth2ParameterNames.REFRESH_TOKEN)) {
            builder.refreshToken((String) tokenResponseParameters.get(OAuth2ParameterNames.REFRESH_TOKEN));
        }

        return builder.build();
    }
}

File: back/src/main/kotlin/com/back/global/security/oauth2/converter/CustomTokenResponseConverter.kt
package com.back.global.security.oauth2.converter

import org.springframework.core.convert.converter.Converter
import org.springframework.security.oauth2.core.OAuth2AccessToken
import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse
import org.springframework.security.oauth2.core.endpoint.OAuth2ParameterNames

class CustomTokenResponseConverter : Converter<Map<String, Any>, OAuth2AccessTokenResponse> {
    override fun convert(tokenResponseParameters: Map<String, Any>): OAuth2AccessTokenResponse {
        val accessToken = tokenResponseParameters[OAuth2ParameterNames.ACCESS_TOKEN] as String
        val expiresIn = tokenResponseParameters[OAuth2ParameterNames.EXPIRES_IN]
            ?.toString()
            ?.toLong()
            ?: 0L
        val scopes = tokenResponseParameters[OAuth2ParameterNames.SCOPE]
            ?.toString()
            ?.split(" ")
            ?.filter { it.isNotBlank() }
            ?.toSet()
            ?: emptySet()

        val additionalParameters = tokenResponseParameters
            .filterKeys { it !in TOKEN_RESPONSE_PARAMETER_NAMES }
            .toMutableMap()

        val refreshTokenObj = tokenResponseParameters[OAuth2ParameterNames.REFRESH_TOKEN]
        if (refreshTokenObj != null) {
            additionalParameters[OAuth2ParameterNames.REFRESH_TOKEN] = refreshTokenObj
        }

        val builder = OAuth2AccessTokenResponse.withToken(accessToken)
            .tokenType(OAuth2AccessToken.TokenType.BEARER)
            .expiresIn(expiresIn)
            .scopes(scopes)
            .additionalParameters(additionalParameters)

        if (refreshTokenObj is String) {
            builder.refreshToken(refreshTokenObj)
        }

        return builder.build()
    }

    companion object {
        private val TOKEN_RESPONSE_PARAMETER_NAMES = setOf(
            OAuth2ParameterNames.ACCESS_TOKEN,
            OAuth2ParameterNames.TOKEN_TYPE,
            OAuth2ParameterNames.EXPIRES_IN,
            OAuth2ParameterNames.REFRESH_TOKEN,
            OAuth2ParameterNames.SCOPE,
        )
    }
}
